[reland]support torch script call over rpc

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32197 [reland]support torch script call over rpc**

This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

zhaojuanmao committed:
[reland]support torch script call over rpc

This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue

ghstack-source-id: 96638829

Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit e015f19e:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e015f19e154ccbf62eba7e5311bc7ff78a984aa4).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4246955) pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build (1/1)
**Step:** "pytorch android gradle build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86194397) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86194397)) :snowflake:
<details>
<summary>
<code>Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:07597f23-fa81-474c-8bef-5c8a91b50595-e015f19e154ccbf62eba7e5311bc7ff78a984aa4-android-arm-v7a not found</code>
</summary>

```
sys	0m0.116s 
++ docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d -w /var/lib/jenkins 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:07597f23-fa81-474c-8bef-5c8a91b50595-e015f19e154ccbf62eba7e5311bc7ff78a984aa4-android-x86_32 
+ export id_x86_32=7b78bb3a17ccdc11b80b147784f71db0017f0774ca857c5f8a83e80d44e6fde2 
+ id_x86_32=7b78bb3a17ccdc11b80b147784f71db0017f0774ca857c5f8a83e80d44e6fde2 
+ export 'COMMAND=((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ COMMAND='((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ echo '((echo' '"source' './workspace/env"' '&&' echo '"sudo' chown -R jenkins 'workspace")' '|' docker exec -u jenkins -i '"$id_x86_32"' 'bash)' '2>&1' 
+ unbuffer bash ./command.sh 
+ ts 
+ docker pull 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:07597f23-fa81-474c-8bef-5c8a91b50595-e015f19e154ccbf62eba7e5311bc7ff78a984aa4-android-arm-v7a 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:07597f23-fa81-474c-8bef-5c8a91b50595-e015f19e154ccbf62eba7e5311bc7ff78a984aa4-android-arm-v7a not found 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zhaojuanmao commented:
the new test failures are real, for both test_script_functions_not_supported and test_backward_simple_script_call, sometimes they timed out "RuntimeError: RPC ran for more than 60000 milliseconds and timed out."

I can not reproduce them locally though, still investigating

zhaojuanmao committed:
Update on "[reland]support torch script call over rpc"


This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

pritamdamania87 commented:
@zhaojuanmao Do you need additional review for any parts of the PR? Or we just need to fix the tests and land this?

zhaojuanmao commented:
so far no need for additional review as no real change to this PR, right now I'm focusing on why there is flakiness, I can find one flaky test among 100 stress runs

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -454,9 +456,32 @@ def rpc_sync(to, func, args=None, kwargs=None):
+    # If invoking an annotated TorchScript function,
+    # call the internal API _rpc_sync_torchscript()
+    if isinstance(func, torch.jit.ScriptFunction):
+        return _rpc_sync_torchscript(to, _qualified_name(func), args, kwargs)
```
Can we put this into `def _invoke_rpc` ?

Also, `def _invoke_rpc` could be renamed to something similar to "dispatch_rpc" to make the functionality of it clearer.
As it dispatches a Python rpc call to one of the 3 code paths.

zhaojuanmao replied:
for autograd profiler, I will address that in follow up diff with more tests, as this diff is already a little bit big:)

zhaojuanmao replied:
tracked in PR #32375

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -424,8 +426,8 @@ def rpc_sync(to, func, args=None, kwargs=None):
         to (str or WorkerInfo): id or name of the destination worker.
-        func (callable): any callable function. builtin functions (like
-                         :meth:`torch.add`) can be sent over RPC more efficiently.
+        func (callable): any callable function. builtin or annotated TorchScript
```
Still feel it's not clear. 

Python function(callable to be precise, which includes lambda too), seems not mentioned from the comment.

```
func (callable): any callable function. builtin or annotated TorchScript
```

to

```
func (callable): any function. Python callable, PyTorch built-in function or annotated TorchScript.
```

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -503,6 +528,146 @@ def rpc_async(to, func, args=None, kwargs=None):
+    """
+    args = args if args else ()
+    kwargs = kwargs if kwargs else {}
+    fut = _invoke_rpc_script(to, qualified_name, *args, **kwargs)
```
Duplicate to me.

```
fut = _rpc_async_torchscript(to, qualified_name, args, kwargs)
```
instead.

xush6528 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -19,6 +19,7 @@ def is_available():
 
 if is_available():
     from .api import _init_rpc_backend, _require_initialized
+    from .api import _rpc_sync_torchscript, _rpc_async_torchscript
```
Is this an unnecessary import? 
Users are not expected to use API starting with underscore, right?

zhaojuanmao replied:
but training toolkit will use it?

xush6528 commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -0,0 +1,55 @@
+  // Script call only allows single IValue returned.
+  auto returns = PythonRpcHandler::getInstance()
+                     .jitCompilationUnit()
+                     ->get_function(qualifiedName)
```
I think you have looked up for this function outside of "rpcTorchscriptCall", can we pass the function as a reference to this function?

zhaojuanmao replied:
rpcTorchscriptCall will be called in JIT later on, so the signature of rpcTorchscriptCall should be compitable with JIT ops, that can not accept function as arg

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -290,6 +292,54 @@ If the future completes with an error, an exception is thrown.
+                            .jitCompilationUnit()
+                            ->get_function(name)
+                            .getSchema();
+        auto stack = torch::jit::createStackForSchema(
```
After all, why do we need to put JIT-related work code in Python binding code?
I believe the python binding code should be thin, and not to bring in implementation details.

zhaojuanmao replied:
we need to make rpcTorchscriptCall() return ivalue::Future, so that it can be called in JIT later on. here other codes are not JIT-related, it is actually python related codes that convert python args to IValues and wrap ivalue future to pyFuture

xush6528 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -290,6 +292,54 @@ If the future completes with an error, an exception is thrown.
+        {
+          // acquiring GIL as torch::jit::toPyObject creates new py::object
+          // without grabbing the GIL.
+          AutoGIL ag;
```
By default, pybind uses acquires GIL when calling C++ functions.

See
https://pybind11.readthedocs.io/en/stable/advanced/misc.html#global-interpreter-lock-gil

So you may want to add a tag, `py::gil_scoped_release`, to increase efficiency of this member function.

zhaojuanmao replied:
that is a good point!

zhaojuanmao replied:
right, looks like it is re-entrant for the same thread. But it could cause deadlock if another thread is trying to acquire the lock and this thread is waiting for another thread to finish. (just got it from debugging the flaky tests:(). So I think it is always a good practice to tag gil release if there are some python stuff called in C++

zhaojuanmao committed:
Update on "[reland]support torch script call over rpc"


This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland]support torch script call over rpc"


This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[reland]support torch script call over rpc"


This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

zhaojuanmao commented:
based on my local debugging, I think I found the root cause for flakiness now...

pythonRpcHandler singleton construction is not thread safe.  it is possible server thread and client thread are constructing pythonRpcHandler at the same time. 

This is not big issue before, because pythonRpcHandler related call only happened in server thread to process call back and in client thread to process result, the chance that the two threads races is very small. 

However, for this diff, we are making  pythonRpcHandler related call very frequently, e.g, call it when we send requests. It has very high chance to have two threads in race state and make the host stuck there. 

The solution is to make pythonRpcHandler thread safe.   (I remember I explicitly constructed it during process group agent and thrift group agent construction so that there will be no races, but it was removed because pure C++ testing environment  did not initialize python interpreter and failed to construct pythonRpcHandler)

zhaojuanmao commented:
The root cause is found, there is deadlock when constructing pythonRpcHandler, the issue is resolved by adding the gil release tag  while pybinding _invoke_rpc_script. 

The deadlock sequence is like this in the stuck host:
1. server thread started to process request, it is trying to wait for GIL to construct python rpc handler
2. client thread started to _invoke_rpc_script() and send request. while calling this from python, pybind hold the GIL
3. inside _invoke_rpc_script(), it is trying to wait for python rpc handler to finish construction
4. server thread is waiting for _invoke_rpc_script() python call to release the GIL

before fixing, 100 stress run can catch some failed tests. after fixing, there is no more failed test in 100 stress run now. 

Let's wait for the CI to double confirm

zhaojuanmao committed:
Update on "[reland]support torch script call over rpc"


This is to reland #30063, the main change is to match a general exception and grep "pickle" error word in "test_script_functions_not_supported" unit test, as Python 3.5 and Python 3.6 throw different types of errors with different error message for the rpc call in the unit test.
[test all]This diff makes following changes:
1. Providing a new set of python rpc privated APIs, they can accept an annotated TorchScript call and this call can be serialized, deserialized and executed in C++ without GIL. These privated APIs will be binded to JIT in the future, and they are different from public APIs as future JIT binded private APIs will be able to accept qualified_name, not callables. These private APIs are subject to be deprecated once JIT supports torch script function to be a JIT type.

Also, these APIs require torch script function to be defined and annotated by users in python land, it can not be script class/module constructor or class/module methods.

2. This diff also allows public rpc APIs to accept an annotated TorchScript call and execute code path that above private APIs ran on. Therefore if users invoke an annotated TorchScript call over RPC, this call can be serialized, deserialized and executed in C++ without GIL as well.

3. The above private APIs call a newly defined C++ function to make rpc torch script call to be serialized, deserialized and executed in C++ land. This C++ function returns an ivalue::Future. so that in follow up diff this C++ function can be called when these privated APIs are binded to JIT.

4. script_call.cpp/.h and request_callback_impl.cpp files are refactored accordingly so that torch script call and builtin call can share same message type and codes.

5. refactored deserializeResponse() and added a new utility to deserizalize response to IValue


Differential Revision: [D19402374](https://our.internmc.facebook.com/intern/diff/D19402374/)

[ghstack-poisoned]

ezyang commented:
This seems to have broken BC:

```
Jan 18 17:52:56 Broken ops: [
Jan 18 17:52:56 	_TorchScriptTesting_StackString::merge(torch.classes._TorchScriptTesting_StackString _0, torch.classes._TorchScriptTesting_StackString _1) -> ()
Jan 18 17:52:56 	_TorchScriptTesting_StackString::clone(torch.classes._TorchScriptTesting_StackString _0) -> (torch.classes._TorchScriptTesting_StackString _0)
Jan 18 17:52:56 ]
```

https://app.circleci.com/jobs/github/pytorch/pytorch/4299484

zhaojuanmao commented:
@ezyang looks like this diff should not have changed PyTorch ops, I'm also not sure where the Broken ops came from.  Who should I contact to confirm whether this is OK or not? thanks

zhaojuanmao commented:
I saw one PR to whitelist some similar jit tests two weeks ago https://github.com/pytorch/pytorch/pull/31949

zhaojuanmao commented:
also the CI backward compatibility test before landing is green somehow

zhaojuanmao commented:
talked with @albanD, the new _TorchScriptTesting_StackString tests were introduced in PR https://github.com/pytorch/pytorch/pull/32320, and they should be whitelisted for back compatibility check because of issue mentioned here https://github.com/pytorch/pytorch/issues/31953. 

I will contact owner of #32320 to add whitelisted ops in BC test.

ezyang commented:
Thanks! I agree with your diagnosis.
